\documentclass{article}
\usepackage{underscore}
\usepackage{booktabs}
\usepackage{float}
\usepackage{subfig}
\usepackage{titling}
\usepackage{fullpage}
\usepackage{pdflscape}

\begin{document}

\title{Data for Figure 2a}
\author{Steve Pederson\\
        Adelaide University\\
        Bioinformatics Centre}
\maketitle

<<loadPackages, include=FALSE, echo=FALSE>>=
library(ggplot2)
library(MASS)
library(matrixStats)
library(vioplot)
library(xtable)
library(multcomp)
library(boot)
source("extraFunctions.R")
@

<<knitrSetup, include=FALSE>>=
knit_hooks$set(crop=hook_pdfcrop)
opts_knit$set(self.contained=FALSE)
opts_chunk$set(fig.align='center', fig.show='as.is', crop=TRUE, include=FALSE, echo=FALSE, cache=TRUE)
Sys.setenv(TEXINPUTS=getwd(), BIBINPUTS=getwd(), BSTINPUTS=getwd())
pdf.options(useDingbats=TRUE)
@


<<loadData>>=
fig2 <- read.csv("Figure2.csv", header=FALSE, stringsAsFactor=FALSE)
colnames(fig2) <- c("Region", "RE", "Tissue", "bpRETE", "bpRE")
fig2$Region <- factor(fig2$Region, levels=c("5'UTR", "Start", "CDS-exon", "CDS-intron", "Stop", "3'UTR", "Intergenic"))
fig2$RE <- factor(fig2$RE, levels=c("Active_promoter","Weak_promoter", "Strong_enhancer", "Weak_enhancer","Repressor", "Insulator"))
levels(fig2$RE)[1:4] <- c("Active Promoter","Weak Promoter", "Strong Enhancer", "Weak Enhancer")
fig2$Tissue <- as.factor(fig2$Tissue)
fig2[,4:5] <- fig2[,4:5]/100 # Convert back to proportions
@

<<collectStats>>=
n <- length(levels(fig2$RE))*length(levels(fig2$Region))
fig2stats <- data.frame(Region = factor(rep(levels(fig2$Region), each=length(levels(fig2$RE))), levels=levels(fig2$Region)),
                       RE = factor(rep(levels(fig2$RE), times=length(levels(fig2$Region))), levels=levels(fig2$RE)),
                       Class = rep(c("BG", "RETE"), each=n))
temp <- matrix(0, nrow=n*2, ncol=3, dimnames=list(c(),c("mean","lower","upper")))
for (i in 1:n){
  x <- logit(fig2$bpRE[intersect(which(fig2$Region==fig2stats$Region[i]), which(fig2$RE==fig2stats$RE[i]))])
  temp[i, 1] <- inv.logit(mean(x))
  temp[i, 2] <- inv.logit(mean(x)-sd(x))
  temp[i, 3] <- inv.logit(mean(x)+sd(x))
  y <- logit(fig2$bpRETE[intersect(which(fig2$Region==fig2stats$Region[i]), which(fig2$RE==fig2stats$RE[i]))])
  temp[i + n, 1] <- inv.logit(mean(y))
  temp[i + n, 2] <- inv.logit(mean(y)-sd(y))
  temp[i + n, 3] <- inv.logit(mean(y)+sd(y))
}
fig2stats <- data.frame(fig2stats, 100*temp)
## Change the lebelling for better reading on facets
levels(fig2stats$Class)[levels(fig2stats$Class)=="BG"] <- "RE in Genomic Region"
levels(fig2stats$Class)[levels(fig2stats$Class)=="RETE"] <- "TE in RE"
@

\section*{Data Preparation}
Data was loaded into R, and the mean \& standard error of the density of bases across all tissues was calculated across all 6 tissues by changing data to proportions and using the logit transformation.
These values were transformed back into percentages for the purposes of plotting in Figure \ref{fig:fig2a}.\\
\\
For the density of Regulatory Elements within each region, and for the density of Transposable Element within each Regulatory Element, data was again analysed using the logit transformation.
A linear regression analysis was performed for the density of Regulatory Elements within each genomic region, and all pairwise comparisons were performed using Tukey's Honestly Significant Difference at the 99\% significance level, after adjustment for multiple comparisons using the R package \textit{multcomp}.
An interval which contained zero was indicative of no difference in the two specific densities being compared.
If an interval was completely positive the first term in the comparison can be considered as enriched in comparison to the other.
Conversely if an interval was exclusively negative the second term in the comparison could be considered as enriched.

\begin{landscape}
<<fig2a, fig.height=7, fig.width=11, include=TRUE, fig.cap='Density of Regulatory Elements shown across multiple genomic regions, denoted in percentage of base pairs (\\%bp), and density of Transposable Elements within Regulatory Elements.'>>=
y_ranges <- data.frame(Region=rep("5'UTR",2), RE=rep("Active Promoter", 2), Class=c("RE in Genomic Region","TE in RE"), mean=c(26, 3.7))
ggplot(fig2stats, aes(x=Region, y=mean, fill=Region)) +
  facet_grid(Class ~ RE, scales="free_y") +
  geom_bar(stat = "identity", position="dodge") + 
  geom_errorbar(aes(ymin=lower, ymax=upper), width=0.2) +
  ylab("Percentage of Base Pairs") + 
  xlab("") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  scale_y_continuous(expand=c(0,0)) + 
  geom_blank(data=y_ranges, aes(x=Region, y=mean))
@
\end{landscape}

\section{The Baseline Densities of Regulatory Elements Within Genomic Regions}

<<setupBGObjects>>=
## The basic approach here will be to perform a series of pairwise comparisons within each region
## Raw p-values for each comparison will not be adjusted until all comparisons have been performed
## Setup an array capable of holding each set of comparisons
comps <- c("Start - 5'UTR", "CDS-exon - 5'UTR", "CDS-intron - 5'UTR", "Stop - 5'UTR", "3'UTR - 5'UTR", "Intergenic - 5'UTR", "CDS-exon - Start", "CDS-intron - Start", "Stop - Start", "3'UTR - Start", "Intergenic - Start", "CDS-intron - CDS-exon", "Stop - CDS-exon", "3'UTR - CDS-exon", "Intergenic - CDS-exon", "Stop - CDS-intron", "3'UTR - CDS-intron", "Intergenic - CDS-intron", "3'UTR - Stop", "Intergenic - Stop", "Intergenic - 3'UTR")
regEls <- levels(fig2stats$RE)
bgComps <- array(0, dim=c(21, 5, 6), dimnames=list(comps, c("coefficients", "sigma", "Tstat", "Pvalues", "adjP"), regEls))
bgList.glht <- vector("list", 6)
names(bgList.glht) <- regEls
for (i in 1:6) {
  bgList.glht[[i]] <- glht(lm(logit(bpRE) ~ 0 + Region, data=subset(fig2, RE==names(bgList.glht)[i])), linfct=mcp(Region="Tukey"))
  bgComps[,1:4,i] <- as.matrix(data.frame(summary(bgList.glht[[i]], test=adjusted(type="none"))$test[3:6]))
}
bgComps[,5,] <- p.adjust(bgComps[,4,], method="bonf")

@
For all genomic regions, the average density of regulatory elements was estimated across all 6 tissues using proportion of basepairs transformed using the logit function (see above).
Pairwise comparisons were then performed for the average densities, and $p$-values were adjusted using Bonferroni's procedure at the family level, i.e. across all regulatory elements in all tissues.

\subsection*{Active Promoters in Genomic Regions}
<<bgAP, include=TRUE, results='asis'>>=
bgAP <- data.frame(Estimate=bgComps[,"coefficients","Active Promoter"], 
                   Tstat=bgComps[,"Tstat","Active Promoter"],
                   PValue = "",
                   Sig = aster(bgComps[,"adjP","Active Promoter"]),
                   stringsAsFactors=FALSE)
bgAP$PValue[which(bgComps[,"adjP","Active Promoter"] > 0.001)] <- format(bgComps[which(bgComps[,"adjP","Active Promoter"] > 0.001),"adjP","Active Promoter"], digits=2)
bgAP$PValue[which(bgComps[,"adjP","Active Promoter"] <= 0.001)] <- format(bgComps[which(bgComps[,"adjP","Active Promoter"] <= 0.001),"adjP","Active Promoter"], digits=3, scientific=TRUE)
bgAP$PValue[which(bgComps[,"adjP","Active Promoter"] == 0)] <- "<2e-16"
print(xtable(bgAP, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:bgAP", 
             caption = "Pairwise comparisons for density of \\textbf{Active Promoters} between genomic regions. \\textit{Performed after logit transformation of the original proportions. A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:bgAP}) revealed that:
\begin{itemize}
\item Unsurprisingly, Active Promoters (APs) are significantly enriched in Start Sites in comparison to all other sampled genomic regions.
\item APs were also enriched in 5'UTRs in comparison to all regions, except CDS-exons \& the aforementioned start sites.
\item CDS-exons were also enriched for APs in comparison to Introns, Stop Sites, 3'UTRs \& Intergenic Regions.
\item Finally, Transcription Stop Sites were marginally enriched for APs in comparison to 3'UTRs \& Intergenic Regions.
\end{itemize}

\clearpage
\subsection*{Weak Promoters in Genomic Regions}
<<bgWP, include=TRUE, results='asis'>>=
bgWP <- data.frame(Estimate=bgComps[,"coefficients","Weak Promoter"], 
                   Tstat=bgComps[,"Tstat","Weak Promoter"],
                   PValue = "",
                   Sig = aster(bgComps[,"adjP","Weak Promoter"]),
                   stringsAsFactors=FALSE)
bgWP$PValue[which(bgComps[,"adjP","Weak Promoter"] > 0.001)] <- format(bgComps[which(bgComps[,"adjP","Weak Promoter"] > 0.001),"adjP","Weak Promoter"], digits=2)
bgWP$PValue[which(bgComps[,"adjP","Weak Promoter"] <= 0.001)] <- format(bgComps[which(bgComps[,"adjP","Weak Promoter"] <= 0.001),"adjP","Weak Promoter"], digits=3, scientific=TRUE)
bgWP$PValue[which(bgComps[,"adjP","Weak Promoter"] == 0)] <- "<2e-16"
print(xtable(bgWP, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:bgWP", 
             caption = "Pairwise comparisons for density of \\textbf{Weak Promoters} between genomic regions. \\textit{Performed after logit transformation of the original proportions. A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:bgWP}) revealed that:
\begin{itemize}
\item Weak Promoters (WPs) are significantly enriched in Start Sites in comparison to all other sampled genomic regions.
\item WPs were also enriched in 5'UTRs in comparison to Introns, 3'UTRs and Intergenic Regions.
\item CDS-exons were also enriched for WPs in comparison to Intergenic Regions only.
\end{itemize}

\clearpage
\subsection*{Strong Enhancers in Genomic Regions}
<<bgSE, include=TRUE, results='asis'>>=
bgSE <- data.frame(Estimate=bgComps[,"coefficients","Strong Enhancer"], 
                   Tstat=bgComps[,"Tstat","Strong Enhancer"],
                   PValue = "",
                   Sig = aster(bgComps[,"adjP","Strong Enhancer"]),
                   stringsAsFactors=FALSE)
bgSE$PValue[which(bgComps[,"adjP","Strong Enhancer"] > 0.001)] <- format(bgComps[which(bgComps[,"adjP","Strong Enhancer"] > 0.001),"adjP","Strong Enhancer"], digits=2)
bgSE$PValue[which(bgComps[,"adjP","Strong Enhancer"] <= 0.001)] <- format(bgComps[which(bgComps[,"adjP","Strong Enhancer"] <= 0.001),"adjP","Strong Enhancer"], digits=3, scientific=TRUE)
bgSE$PValue[which(bgComps[,"adjP","Strong Enhancer"] == 0)] <- "<2e-16"
print(xtable(bgSE, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:bgSE", 
             caption = "Pairwise comparisons for density of \\textbf{Strong Enhancers} between genomic regions. \\textit{Performed after logit transformation of the original proportions. A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:bgSE}) revealed that:
\begin{itemize}
\item Strong Enhancers (SEs) are significantly enriched in only in 5'UTRs compared to Exons, 3'UTRs \& Intergenic Regions
\end{itemize}

\clearpage
\subsection*{Weak Enhancers in Genomic Regions}
<<bgWE, include=TRUE, results='asis'>>=
bgWE <- data.frame(Estimate=bgComps[,"coefficients","Weak Enhancer"], 
                   Tstat=bgComps[,"Tstat","Weak Enhancer"],
                   PValue = "",
                   Sig = aster(bgComps[,"adjP","Weak Enhancer"]),
                   stringsAsFactors=FALSE)
bgWE$PValue[which(bgComps[,"adjP","Weak Enhancer"] > 0.001)] <- format(bgComps[which(bgComps[,"adjP","Weak Enhancer"] > 0.001),"adjP","Weak Enhancer"], digits=2)
bgWE$PValue[which(bgComps[,"adjP","Weak Enhancer"] <= 0.001)] <- format(bgComps[which(bgComps[,"adjP","Weak Enhancer"] <= 0.001),"adjP","Weak Enhancer"], digits=3, scientific=TRUE)
bgWE$PValue[which(bgComps[,"adjP","Weak Enhancer"] == 0)] <- "<2e-16"
print(xtable(bgWE, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:bgWE", 
             caption = "Pairwise comparisons for density of \\textbf{Weak Enhancers} between genomic regions. \\textit{Performed after logit transformation of the original proportions. A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:bgWE}) revealed that:
\begin{itemize}
\item Weak Enhancers (WEs) are significantly enriched in 5'UTRs in comparison to Exons, Stop Sites \& 3'UTRs
\item WEs are enriched in Start Sites, Introns \& Intergenic regions compared to Exons.
\end{itemize}

\subsection*{Repressors and Insulators in Genomic Regions}
<<bgRP, include=FALSE>>=
bgRP <- data.frame(Estimate=bgComps[,"coefficients","Repressor"], 
                   Tstat=bgComps[,"Tstat","Repressor"],
                   PValue = "",
                   Sig = aster(bgComps[,"adjP","Repressor"]),
                   stringsAsFactors=FALSE)
bgRP$PValue[which(bgComps[,"adjP","Repressor"] > 0.001)] <- format(bgComps[which(bgComps[,"adjP","Repressor"] > 0.001),"adjP","Repressor"], digits=2)
bgRP$PValue[which(bgComps[,"adjP","Repressor"] <= 0.001)] <- format(bgComps[which(bgComps[,"adjP","Repressor"] <= 0.001),"adjP","Repressor"], digits=3, scientific=TRUE)
bgRP$PValue[which(bgComps[,"adjP","Repressor"] == 0)] <- "<2e-16"
print(xtable(bgRP, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:bgRP", 
             caption = "Pairwise comparisons for density of \\textbf{Repressors} between genomic regions. \\textit{Performed after logit transformation of the original proportions. A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
<<bgIN, include=FALSE>>=
bgIN <- data.frame(Estimate=bgComps[,"coefficients","Insulator"], 
                   Tstat=bgComps[,"Tstat","Insulator"],
                   PValue = "",
                   Sig = aster(bgComps[,"adjP","Insulator"]),
                   stringsAsFactors=FALSE)
bgIN$PValue[which(bgComps[,"adjP","Insulator"] > 0.001)] <- format(bgComps[which(bgComps[,"adjP","Insulator"] > 0.001),"adjP","Insulator"], digits=2)
bgIN$PValue[which(bgComps[,"adjP","Insulator"] <= 0.001)] <- format(bgComps[which(bgComps[,"adjP","Insulator"] <= 0.001),"adjP","Insulator"], digits=3, scientific=TRUE)
bgIN$PValue[which(bgComps[,"adjP","Insulator"] == 0)] <- "<2e-16"
print(xtable(bgIN, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:bgIN", 
             caption = "Pairwise comparisons for density of \\textbf{Insulators} between genomic regions. \\textit{Performed after logit transformation of the original proportions. A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@

No differences in the distributions of Repressors or Insulators was revealed after pairwise comparisons (all adjusted $p$-values=1).

\clearpage
\section{The Densities of Transposable Elements Within Regulatory Elements Across Genomic Regions}
<<setupREObjects>>=
## The basic approach here will be to perform a series of pairwise comparisons within each region
## Raw p-values for each comparison will not be adjusted until all comparisons have been performed
## Setup an array capable of holding each set of comparisons
teComps <- array(0, dim=c(21, 5, 6), dimnames=list(comps, c("coefficients", "sigma", "Tstat", "Pvalues", "adjP"), regEls))
teList.glht <- vector("list", 6)
names(teList.glht) <- regEls
for (i in 1:6) {
  teList.glht[[i]] <- glht(lm(logit(bpRETE) ~ 0 + Region, data=subset(fig2, RE==names(teList.glht)[i])), linfct=mcp(Region="Tukey"))
  teComps[,1:4,i] <- as.matrix(data.frame(summary(teList.glht[[i]], test=adjusted(type="none"))$test[3:6]))
}
teComps[,5,] <- p.adjust(teComps[,4,], method="bonf")
@
The same approach as for Regulatory Elements within regions was applied for Transposable Elements within Regulatory Elements (i.e. RETEs) across all sampled genomic regions.
The Bonferroni procedure was also applied to all $p$-values after all comparisons had been performed.

\subsection*{Transposable Elements Across Active Promoters}
<<teAP, include=TRUE, results='asis'>>=
teAP <- data.frame(Estimate=teComps[,"coefficients","Active Promoter"], 
                   Tstat=teComps[,"Tstat","Active Promoter"],
                   PValue = "",
                   Sig = aster(teComps[,"adjP","Active Promoter"]),
                   stringsAsFactors=FALSE)
teAP$PValue[which(teComps[,"adjP","Active Promoter"] > 0.001)] <- format(teComps[which(teComps[,"adjP","Active Promoter"] > 0.001),"adjP","Active Promoter"], digits=2)
teAP$PValue[which(teComps[,"adjP","Active Promoter"] <= 0.001)] <- format(teComps[which(teComps[,"adjP","Active Promoter"] <= 0.001),"adjP","Active Promoter"], digits=3, scientific=TRUE)
teAP$PValue[which(teComps[,"adjP","Active Promoter"] == 0)] <- "<2e-16"
print(xtable(teAP, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:teAP", 
             caption = "Pairwise comparisons for density of \\textbf{TEs} within \\textbf{Active Promoters} between genomic regions. \\textit{A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:teAP}) revealed that:
\begin{itemize}
\item Active Promoters within 5'UTRs and Start Sites showed a greater density of TEs than APs within all other Genomic Regions, with the exception of the comparison against each other
\item APs within Exons, Introns showed a greater density of TEs when compared to Stop Sites \& 3'UTRs
\item APs within Intergenic Regions showed a greater density of TEs in comparison to Stop Sites only.
\end{itemize}

\clearpage
\subsection*{Transposable Elements Across Weak Promoters}
<<teWP, include=TRUE, results='asis'>>=
teWP <- data.frame(Estimate=teComps[,"coefficients","Weak Promoter"], 
                   Tstat=teComps[,"Tstat","Weak Promoter"],
                   PValue = "",
                   Sig = aster(teComps[,"adjP","Weak Promoter"]),
                   stringsAsFactors=FALSE)
teWP$PValue[which(teComps[,"adjP","Weak Promoter"] > 0.001)] <- format(teComps[which(teComps[,"adjP","Weak Promoter"] > 0.001),"adjP","Weak Promoter"], digits=2)
teWP$PValue[which(teComps[,"adjP","Weak Promoter"] <= 0.001)] <- format(teComps[which(teComps[,"adjP","Weak Promoter"] <= 0.001),"adjP","Weak Promoter"], digits=3, scientific=TRUE)
teWP$PValue[which(teComps[,"adjP","Weak Promoter"] == 0)] <- "<2e-16"
print(xtable(teWP, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:teWP", 
             caption = "Pairwise comparisons for density of \\textbf{TEs} within \\textbf{Weak Promoters} between genomic regions. \\textit{A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:teWP}) revealed that:
\begin{itemize}
\item Weak Promoters within 5'UTRs showed a greater density of TEs than WPs within all other Genomic Regions, including the comparison against Start Sites
\item WPs showed a greater density in Start Sites in comparison to Exons, Stop Sites \& 3'UTRs only
\item WPs within Exons, Introns \& Intergenic Regions all showed a greater density of TEs when compared to Stop Sites
\item WPs within 3'UTRs showed an increased density of TEs in comparison to Stop Site.
\end{itemize}

\clearpage
\subsection*{Transposable Elements Across Strong Enhancers}
<<teSE, include=TRUE, results='asis'>>=
teSE <- data.frame(Estimate=teComps[,"coefficients","Strong Enhancer"], 
                   Tstat=teComps[,"Tstat","Strong Enhancer"],
                   PValue = "",
                   Sig = aster(teComps[,"adjP","Strong Enhancer"]),
                   stringsAsFactors=FALSE)
teSE$PValue[which(teComps[,"adjP","Strong Enhancer"] > 0.001)] <- format(teComps[which(teComps[,"adjP","Strong Enhancer"] > 0.001),"adjP","Strong Enhancer"], digits=2)
teSE$PValue[which(teComps[,"adjP","Strong Enhancer"] <= 0.001)] <- format(teComps[which(teComps[,"adjP","Strong Enhancer"] <= 0.001),"adjP","Strong Enhancer"], digits=3, scientific=TRUE)
teSE$PValue[which(teComps[,"adjP","Strong Enhancer"] == 0)] <- "<2e-16"
print(xtable(teSE, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:teSE", 
             caption = "Pairwise comparisons for density of \\textbf{TEs} within \\textbf{Strong Enhancers} between genomic regions. \\textit{A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:teSE}) revealed that:
\begin{itemize}
\item Strong Enhancers (SEs) within 5'UTRs showed a greater density of TEs than SEs within all other Genomic Regions except Introns
\item SEs showed a greater density in Start Sites in comparison to Exons only
\item SEs within Introns showed a greater density in comparison to Start Sites, Exons, Stop Sites \& 3'UTRs
\item SEs within 3'UTRs showed a greater density in comparison to Start Sites, Exons \& Stop Sites
\item SEs within Intergenic Regions showed a greater density of TEs in comparison to Start Sites, Exons \& Stop Sites
\end{itemize}

\clearpage
\subsection*{Transposable Elements Across Weak Enhancers}
<<teWE, include=TRUE, results='asis'>>=
teWE <- data.frame(Estimate=teComps[,"coefficients","Weak Enhancer"], 
                   Tstat=teComps[,"Tstat","Weak Enhancer"],
                   PValue = "",
                   Sig = aster(teComps[,"adjP","Weak Enhancer"]),
                   stringsAsFactors=FALSE)
teWE$PValue[which(teComps[,"adjP","Weak Enhancer"] > 0.001)] <- format(teComps[which(teComps[,"adjP","Weak Enhancer"] > 0.001),"adjP","Weak Enhancer"], digits=2)
teWE$PValue[which(teComps[,"adjP","Weak Enhancer"] <= 0.001)] <- format(teComps[which(teComps[,"adjP","Weak Enhancer"] <= 0.001),"adjP","Weak Enhancer"], digits=3, scientific=TRUE)
teWE$PValue[which(teComps[,"adjP","Weak Enhancer"] == 0)] <- "<2e-16"
print(xtable(teWE, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:teWE", 
             caption = "Pairwise comparisons for density of \\textbf{TEs} within \\textbf{Weak Enhancers} between genomic regions. \\textit{A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:teWE}) revealed that:
\begin{itemize}
\item Weak Enhancers (WEs) within 5'UTRs showed a greater density of TEs than WEs within all other Genomic Regions except Introns \& Intergenic Regions
\item WEs showed a greater density in Start Sites in comparison to Exons \& Stop Sites
\item WEs within Introns showed a greater density in comparison to Start Sites, Exons, Stop Sites \& 3'UTRs
\item WEs within 3'UTRs showed a greater density in comparison to Start Sites, Exons \& Stop Sites
\item WEs within Intergenic Regions showed a greater density of TEs in comparison to Start Sites, Exons \& Stop Sites
\end{itemize}

\clearpage
\subsection*{Transposable Elements Across Repressors}
<<teRP, include=TRUE, results='asis'>>=
teRP <- data.frame(Estimate=teComps[,"coefficients","Repressor"], 
                   Tstat=teComps[,"Tstat","Repressor"],
                   PValue = "",
                   Sig = aster(teComps[,"adjP","Repressor"]),
                   stringsAsFactors=FALSE)
teRP$PValue[which(teComps[,"adjP","Repressor"] > 0.001)] <- format(teComps[which(teComps[,"adjP","Repressor"] > 0.001),"adjP","Repressor"], digits=2)
teRP$PValue[which(teComps[,"adjP","Repressor"] <= 0.001)] <- format(teComps[which(teComps[,"adjP","Repressor"] <= 0.001),"adjP","Repressor"], digits=3, scientific=TRUE)
teRP$PValue[which(teComps[,"adjP","Repressor"] == 0)] <- "<2e-16"
print(xtable(teRP, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:teRP", 
             caption = "Pairwise comparisons for density of \\textbf{TEs} within \\textbf{Repressors} between genomic regions. \\textit{A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:teRP}) revealed that:
\begin{itemize}
\item Repressors (RPs) within 5'UTRs showed a greater density of TEs than RPs within Start Sites, Exons \& Stop Sites
\item RPs within Introns showed a greater density in comparison to Start Sites, Exons \& Stop Sites
\item RPs within 3'UTRs showed a greater density in comparison to Start Sites, Exons \& Stop Sites
\item RPs within Intergenic Regions showed a greater density of TEs in comparison to Start Sites, Exons \& Stop Sites
\end{itemize}

\clearpage
\subsection*{Transposable Elements Across Insulators}
<<teIN, include=TRUE, results='asis'>>=
teIN <- data.frame(Estimate=teComps[,"coefficients","Insulator"], 
                   Tstat=teComps[,"Tstat","Insulator"],
                   PValue = "",
                   Sig = aster(teComps[,"adjP","Insulator"]),
                   stringsAsFactors=FALSE)
teIN$PValue[which(teComps[,"adjP","Insulator"] > 0.001)] <- format(teComps[which(teComps[,"adjP","Insulator"] > 0.001),"adjP","Insulator"], digits=2)
teIN$PValue[which(teComps[,"adjP","Insulator"] <= 0.001)] <- format(teComps[which(teComps[,"adjP","Insulator"] <= 0.001),"adjP","Insulator"], digits=3, scientific=TRUE)
teIN$PValue[which(teComps[,"adjP","Insulator"] == 0)] <- "<2e-16"
print(xtable(teIN, 
             digits = c(0,3,3,0,0), 
             align = c("r","r","r","r","l"),
             label = "tab:teIN", 
             caption = "Pairwise comparisons for density of \\textbf{TEs} within \\textbf{Insulators} between genomic regions. \\textit{A positive value for the Estimate equates to a higher density in the first genomic region, whereas a negative value equates to a higher density in the second genomic region.}"), 
      caption.placement="top")
@
Pairwise comparisons (Table \ref{tab:teIN}) revealed that:
\begin{itemize}
\item Insulators (INs) within 5'UTRs showed a greater density of TEs than INs within Start Sites, Exons \& Stop Sites
\item INs within Introns showed a greater density in comparison to Start Sites, Exons \& Stop Sites
\item INs within 3'UTRs showed a greater density in comparison to Start Sites, Exons \& Stop Sites
\item INs within Intergenic Regions showed a greater density of TEs in comparison to Start Sites, Exons \& Stop Sites
\end{itemize}

\section{Fig 2a - Conclusion}
Amongst the avalanche of data presented above, the striking pattern became evident that non-coding regions showed a greater density of Transposable Elements within many Regulatory Elements, most notably for Insulators, Repressors \& Enhancers.
In general, there was also minimal significant difference in the densities of TEs across any of the coding regions.

%------------------------------------
% Also compare across tissues
%------------------------------------

\end{document}