\documentclass{article}
\usepackage{underscore}
\usepackage{booktabs}
\usepackage{float}
\usepackage{subfig}
\usepackage{titling}
\usepackage{fullpage}
\usepackage{pdflscape}
\usepackage{placeins}

\begin{document}

\title{Data for Figure 2}
\author{Steve Pederson\\
        Adelaide University\\
        Bioinformatics Centre}
\maketitle

<<loadPackages, message=FALSE, warning=FALSE>>=
library(ggplot2)
library(MASS)
library(matrixStats)
library(xtable)
library(multcomp)
library(sandwich)
library(car)
library(boot)
library(dplyr)
library(readr)
library(knitr)
library(magrittr)
library(limma)
source("extraFunctions.R")
@

<<knitrSetup, include=FALSE>>=
knit_hooks$set(crop=hook_pdfcrop)
opts_knit$set(self.contained=FALSE)
opts_chunk$set(fig.align='center', fig.show='as.is', crop=TRUE, 
               include=TRUE, echo=TRUE, results = 'hide',
               warning = FALSE, message = FALSE)
Sys.setenv(TEXINPUTS=getwd(), BIBINPUTS=getwd(), BSTINPUTS=getwd())
pdf.options(useDingbats=TRUE)
@

\section{Loading \& Transforming the Data}
<<loadData>>=
fig2 <- read_csv("Figure2.csv", col_names = FALSE) %>%
  set_colnames(c("Region", "RE", "Tissue", "bpRETE", "bpRE")) %>%
  #filter(!Region %in% c("Start", "Stop")) %>%
  mutate(Region = gsub("([35])'UTR", "UTR\\1", Region),
         Region = gsub("CDS-", "CDS", Region),
         Region = factor(Region, levels=c("UTR5", "CDSexon", "CDSintron",  
                                          "UTR3", "Intergenic")),
         RE = gsub("Repressor", "Repressed_region", RE),
         RE = factor(RE, levels= c("Active_promoter","Weak_promoter", "Strong_enhancer", 
                                   "Weak_enhancer", "Repressed_region", "Insulator")),
         Tissue = as.factor(Tissue),
         bpRETE = bpRETE / 100,
         bpRE = bpRE/100,
         logitRETE = logit(bpRETE),
         logitRE = logit(bpRE)) 
@

\begin{itemize}
  \item The raw data was loaded, omitting the Start \& Stop Regions
  \item To ensure valid syntactic names for downstream analysis, the prime symbol (') and the dash (-) were removed from Region names.
  Spaces were converted to underscores.
  \item Percentages of base pairs found to be REs or RETEs were converted to proportions and transformed using the logit transformation to ensure approximate normality
  \item The logit-transformed data were plotted for both REs (Figure \ref{fig:plotRawRE}) and RETEs (Figure \ref{fig:plotRawRETE}) to ensure similar distributions within each Regulatory Element
\end{itemize}

<<plotRawRE, echo=FALSE, fig.height=4, fig.cap='Logit transformed proportions across all 6 tissues for Regulatory Elements'>>=
fig2 %>% 
  ggplot(aes(x = Region, y = logitRE, fill = Region)) + 
  geom_boxplot() + 
  facet_wrap(~RE) +
  theme_bw() +
  labs(y = expression(paste("logit(",pi[RE], ")"))) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
@


<<plotRawRETE, echo=FALSE, fig.height=4, fig.cap='Logit transformed proportions across all 6 tissues for Regulatory Elements with Transposable Elements'>>=
fig2 %>% 
  ggplot(aes(x = Region, y = logitRETE, fill = Region)) + 
  geom_boxplot() + 
  facet_wrap(~RE) +
  theme_bw() +
  labs(y = expression(paste("logit(",pi[RETE], ")"))) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
@

\FloatBarrier
\section{Analysis of Regulatory Elements}

The approach taken is to model the probability of each base within a region belonging to a Regulatory Element ($\pi_i: i = 1, 2, \ldots, 6$), as a proxy for the density of these elements within each region.
As the REs are mutually exclusive, these probabilities are not independent and should be modelled separately, but can be modelled across all genomic regions as these stretches of the genome can be considered to be independent.
The samples from the 6 different tissues are considered to be biological replicates indicative of an average measure across all tissues, with the proportion of bases belonging to each RE class being representative of the probability $\pi$.\\

\subsection{Initial Model Checks}
The first thing to assess is whether the variance between the replicates is approximately equal between regions.
This test for homoscedasticity was performed using Levene's Test, with results given in Table \ref{tab:leveneRE}.
Levene's test is a test for the null hypothesis $H_0: \sigma_1 = \sigma_2 = \ldots = \sigma_6$, with the alternative that the variances are not all equal.
No evidence of the inequality of variances was detected within any region, and this gives initial support to being able to use linear regression on this data. \\
\\
A linear regression model was then fit for each separate RE, and Normal QQ plots were inspected for approximate normality (Figure \ref{fig:qqNormRE}).
Despite a few outlier points being evident, no strong evidence for violation of this assumption were detected, and any subsequent results may be considered meaningful.

<<leveneRE, warning=FALSE, echo=FALSE, include=TRUE, results='asis'>>=
fig2 %>%
  extract2("RE") %>%
  levels() %>%
  lapply(function(x){data.frame(RE = x,
                                leveneTest(logitRE ~ Region, 
                                data = filter(fig2, RE == x), 
                                center = mean)[1,])}) %>%
  bind_rows %>%
  set_colnames(c("Regulatory Element", "DF", "F Statistic", "P Value")) %>%
  xtable(label = "tab:leveneRE",
         caption = paste("Levene's Test for homogeneity of variance across regions within each RE.",
                         "The null hypothesis was not rejected in any RE, ",
                         "thus we can conclude that variances are equal across regions"),
         digits = c(0, 0, 0, 3, 3),
         align = "llrrr") %>%
  print(include.rownames = FALSE,
        caption.placement = "top")
@


<<lmRE>>=
lmRE <- fig2 %>%
  extract2("RE") %>%
  levels() %>%
  sapply(function(x){
    lm(logitRE ~ 0 +Region, data = filter(fig2, RE == x))
  },
  simplify = FALSE)
@


<<qqNormRE, echo = FALSE, fig.height=6, fig.cap='Normal QQ plots for the residuals from each Regulatory Element.'>>=
par(mfrow = c(2, 3))
names(lmRE) %>%
  lapply(function(x){
         plot(lmRE[[x]],
          which = 2,
          caption = NA,
          main = paste0(x, ":\n Normal Q-Q"))
  })
par(mfrow = c(1, 1))
@

\FloatBarrier
\subsection{Results of Analysis for Regulatory Elements}
The test being used to detect any differing probabilities across the genomic regions will be Tukey's Honestly Significant Difference (HSD), which depends on the assumptions of normality and homoscedasticity performed above.
Comparisons across regions will be performed separately for each RE, and the $p$-values for the complete set of REs will be adjusted for multiple comparisons using Bonferroni's methods to control the FWER across the entire experiment.
All results from pairwise comparisons are presented in Tables \ref{tab:REActivepromoter} to \ref{tab:REInsulator}.

<<resRE>>=
resRE <- names(lmRE) %>%
  lapply(function(x){
    glht(lmRE[[x]], linfct = mcp(Region = "Tukey")) %>%
      summary(adjusted(type = "none")) %>%
      extract2("test") %>%
      extract(c("coefficients", "sigma", "tstat", "pvalues")) %>%
      as.data.frame %>%
      add_rownames("Comparison") %>%
      mutate(RE = x)}) %>%
  bind_rows %>%
  dplyr::select(RE, Comparison, 
                Estimate = coefficients,
                Std.Error = sigma,
                T = tstat,
                P = pvalues) %>%
  mutate(adjP = p.adjust(P, "bonferroni"),
         Sig = aster(adjP),
         Comparison = gsub("UTR(3|5)", "\\1'UTR", Comparison))
@

<<printResRE, echo=FALSE, include=TRUE, results='asis'>>=
xtabRE <- invisible(levels(fig2$RE) %>%
  lapply(function(x){
    resRE %>%
      filter(RE == x) %>%
      dplyr::select(Comparison, Estimate, 
                    `T-statistic` = T, 
                    `p-value` = adjP, 
                    ` ` = Sig) %>%
      xtable(label = gsub("_", "", paste0("tab:RE", x)),
             digits = c(0, 0, 3, 3, -2, 0),
             align = "llrrrl",
             caption = paste("Pairwise comparisons for density of \\textbf{", 
                             capwords(gsub("_", " ", x)),
                             "Elements} between genomic regions.",
                             "Pairwise comparisons were performed using Tukey's HSD on the logit-transformed proportions of bases derived from the six sampled tissues.",
                             "The FWER was controlled using Bonferroni's adjustment on the set of $p$-values across all elements, and all $p$-values presented have been adjusted by this method.",
                             "A positive value in the Estimate column corresponds to a higher proportion in the first of the two genomic regions within the comparisons, whilst a negative value corresponds to a higher proportion in the second of the two regions.",
                             "Any significant values are indicated using asterisks as per the standard conventions of R.")) 
  }) %>%
    lapply(print,
           include.rownames = FALSE,
           caption.placement = "top",
           print.results = FALSE) %>%
  unlist %>%
  (function(x){gsub("(T|p)-", "$\\1$-",x)}))
cat(xtabRE)
@

\FloatBarrier
\section{Analysis of Transposable Elements Within Regulatory Elements}

The same approach as above was used to assess any differing probabilities of a base within a given region being a TE within an RE, i.e. RETE.

\subsection{Initial Model Checks}
Levene's test was again used to assess homoscedasticity between regions for each RE.
In contrast to this test for REs alone, evidence of heteroscedasticity was detected in the samples for Active Promoter regions (Table \ref{tab:leveneRETE}; $p = 0.021$).
Inspection of the raw data in Figure \ref{fig:plotRawRETE} revealed lower than usual variance between measures in the CDS-exon RETE values and this is the likely source of this significant result.
Weak evidence was also suggestive of heteroscedasticity for Weak Enhancer samples ($p = 0.074$).
To counteract these findings, the sandwich estimator was used for the variance-covariance matrix in pairwise comparisons for all RETEs.
\\
A linear regression model was then fit for each separate RETE, and Normal QQ plots were inspected for approximate normality (Figure \ref{fig:qqNormRETE}).
Once again, despite a few outlier points being evident, no strong evidence for violation of this assumption were detected, and any subsequent results may be considered meaningful.

<<leveneRETE, warning=FALSE, echo=FALSE, include=TRUE, results='asis'>>=
fig2 %>%
  extract2("RE") %>%
  levels() %>%
  lapply(function(x){data.frame(RE = x,
                                leveneTest(logitRETE ~ Region, 
                                data = filter(fig2, RE == x), 
                                center = mean)[1,])}) %>%
  bind_rows %>%
  set_colnames(c("Regulatory Element", "DF", "F Statistic", "P Value")) %>%
    mutate(` ` = aster(`P Value`)) %>%
  xtable(label = "tab:leveneRETE",
         caption = paste("Levene's Test for homogeneity of variance across regions within each RE.",
                         "The null hypothesis was rejected in Active Promoters,",
                         "with further weak evidence seen in Weak Enhancers."),
         digits = c(0, 0, 0, 3, 3, 0),
         align = "llrrrl") %>%
  print(include.rownames = FALSE,
        caption.placement = "top")
@


<<lmRETE>>=
lmRETE <- fig2 %>%
  extract2("RE") %>%
  levels() %>%
  sapply(function(x){
    lm(logitRETE ~ 0 + Region, data = filter(fig2, RE == x))
  },
  simplify = FALSE)
@


<<qqNormRETE, echo = FALSE, fig.height=6, fig.cap='Normal QQ plots for the residuals from each Regulatory Element.'>>=
par(mfrow = c(2, 3))
names(lmRETE) %>%
  lapply(function(x){
         plot(lmRETE[[x]],
          which = 2,
          caption = NA,
          main = paste0(x, ":\n Normal Q-Q"))
  })
par(mfrow = c(1, 1))
@

\FloatBarrier
\subsection{Results of Analysis for Transposable Elements Within Regulatory Elements}
The test being used to detect any differing probabilities across the genomic regions will again be Tukey's Honestly Significant Difference (HSD), which depends on the assumptions of normality and homoscedasticity performed above.
As also mentioned above, the sandwich estimator for the variance-covariance matrix was used in all pairwise comparisons to counteract any effects due to heteroscedasticity. \\
\\
Comparisons across regions will be performed separately for each RETE, and the $p$-values for the complete set of RETEs will be adjusted for multiple comparisons using Bonferroni's methods to control the FWER across the entire experiment.
All results from pairwise comparisons are presented in Tables \ref{tab:RETEActivepromoter} to \ref{tab:RETEInsulator}.

<<resRETE>>=
resRETE <- names(lmRETE) %>%
  lapply(function(x){
    glht(lmRETE[[x]], linfct = mcp(Region = "Tukey"), vcov = sandwich) %>%
      summary(adjusted(type = "none")) %>%
      extract2("test") %>%
      extract(c("coefficients", "sigma", "tstat", "pvalues")) %>%
      as.data.frame %>%
      add_rownames("Comparison") %>%
      mutate(RE = x)}) %>%
  bind_rows %>%
  dplyr::select(RE, Comparison, 
                Estimate = coefficients,
                Std.Error = sigma,
                T = tstat,
                P = pvalues) %>%
  mutate(adjP = p.adjust(P, "bonferroni"),
         Sig = aster(adjP),
         Comparison = gsub("UTR(3|5)", "\\1'UTR", Comparison))
@

<<printResRETE, echo=FALSE, include=TRUE, results='asis'>>=
xtabRETE <- invisible(levels(fig2$RE) %>%
  lapply(function(x){
    resRETE %>%
      filter(RE == x) %>%
      dplyr::select(Comparison, Estimate, 
                    `T-statistic` = T, 
                    `p-value` = adjP, 
                    ` ` = Sig) %>%
      xtable(label = gsub("_", "", paste0("tab:RETE", x)),
             digits = c(0, 0, 3, 3, -2, 0),
             align = "llrrrl",
             caption = paste("Pairwise comparisons for density of TEs within \\textbf{", 
                             capwords(gsub("_", " ", x)),
                             "Elements} between genomic regions.",
                             "Pairwise comparisons were performed using Tukey's HSD on the logit-transformed proportions of bases derived from the six sampled tissues.",
                             "The FWER was controlled using Bonferroni's adjustment on the set of $p$-values across all elements, and all $p$-values presented have been adjusted by this method.",
                             "A positive value in the Estimate column corresponds to a higher proportion in the first of the two genomic regions within the comparisons, whilst a negative value corresponds to a higher proportion in the second of the two regions.",
                             "Any significant values are indicated using asterisks as per the standard conventions of R.")) 
  }) %>%
    lapply(print,
           include.rownames = FALSE,
           caption.placement = "top",
           print.results = FALSE) %>%
  unlist %>%
  (function(x){gsub("(T|p)-", "$\\1$-",x)}))
cat(xtabRETE)
@


<<paperFig2, echo = FALSE, fig.height=4, fig.cap = 'Figure 2 as it currently appears in the paper'>>=
fs <- 8
c(lmRE, lmRETE) %>% 
  lapply(summary) %>% 
  lapply(coef) %>%
  lapply(as.data.frame) %>%
  lapply(add_rownames, var = "Region") %>%
  bind_rows %>%
  mutate(RE = rep(rep(names(lmRE), each = 5), 2),
         RE = gsub("_", " ", RE),
         RE = capwords(RE),
         #RE = gsub(" ", "\n", RE),
         RE = factor(RE, levels = unique(RE)),
         Region = gsub("Region", "", Region),
         Region = gsub("UTR(3|5)", "\\1'UTR", Region),
         Region = gsub("CDS", "CDS ", Region),
         Region = factor(Region, levels = unique(Region)),
         Class = rep(c("RE in Genomic Region", "RE with TEs"), each = 30)) %>%
  mutate(pi = binomial()$linkinv(Estimate),
         lower = binomial()$linkinv(Estimate - `Std. Error`),
         upper = binomial()$linkinv(Estimate + `Std. Error`)) %>%
  ggplot(aes(x = Region, y = pi, fill = Region)) + 
  geom_bar(stat = "identity") +
  geom_errorbar(aes(ymin = lower, ymax = upper),
                width= 0.4) +
  theme_bw() +
  labs(x = "", y = expression(hat(pi))) +
  scale_y_continuous(expand=c(0.04,0)) + 
  facet_grid(Class~RE, scales = "free_y") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = 0, size = fs),
        axis.text.y = element_text(size = fs-1),
        strip.text = element_text(size = fs-1),
        legend.text = element_text(size = fs),
        legend.title = element_text(size = fs),
        panel.margin = unit(.004, "npc"),
        panel.grid = element_line(colour = "grey50")) 
@



\FloatBarrier
\section{Analysis of Transposable Elements Given Repeat Elements}
One further exploration not undertaken before is the analysis of the conditional probabilities for a base being a TE, \textit{given that the base belongs to a regulatory element}, i.e. $\pi = p(TE|RE)$
The raw data was plotted in Figure \ref{fig:plotTEgivRE} with some potentially interesting patterns being clearly evident.

<<addConditionalTerm>>=
fig2 %<>% 
  mutate(bpTEgivRE = bpRETE / bpRE,
         logitTEgivRE = logit(bpTEgivRE))
@


<<plotTEgivRE, echo=FALSE, fig.height=4, fig.cap='Logit transformed proportions across all 6 tissues for the probability of a base being a Transposable Element, given that the base is in a Regulatory Element'>>=
fig2 %>% 
  ggplot(aes(x = Region, y = logitTEgivRE, fill = Region)) + 
  geom_boxplot() + 
  facet_wrap(~RE) +
  theme_bw() +
  labs(y = expression(paste("logit(",pi[TE], "|", pi[RE], ")"))) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
@

\subsection{Initial Model Checks}
Levene's test was once again used to assess homoscedasticity between regions for each measurement of $p(TE | RE)$.
Evidence of heteroscedasticity was again detected in the samples for Active Promoter regions (Table \ref{tab:leveneTEgivRE}; $p = 0.002$).
Inspection of the raw data in Figure \ref{fig:plotTEgivRE} revealed lower than usual variance between measures in the CDS-exon RETE values and this is the likely source of this significant result.
Weak evidence was also suggestive of heteroscedasticity for Repressed Region samples ($p = 0.067$).
To counteract these findings, the sandwich estimator was used for the variance-covariance matrix in pairwise comparisons for all RETEs.
\\
A linear regression model was then fit for each separate RETE, and Normal QQ plots were inspected for approximate normality (Figure \ref{fig:qqNormTEgivRE}).
Once again, despite a few outlier points being evident, no strong evidence for violation of this assumption were detected, and any subsequent results may be considered meaningful.

<<leveneTEgivRE, warning=FALSE, echo=FALSE, include=TRUE, results='asis'>>=
fig2 %>%
  extract2("RE") %>%
  levels() %>%
  lapply(function(x){data.frame(RE = x,
                                leveneTest(logitTEgivRE ~ Region, 
                                data = filter(fig2, RE == x), 
                                center = mean)[1,])}) %>%
  bind_rows %>%
  set_colnames(c("Regulatory Element", "DF", "F Statistic", "P Value")) %>%
    mutate(` ` = aster(`P Value`)) %>%
  xtable(label = "tab:leveneTEgivRE",
         caption = paste("Levene's Test for homogeneity of variance across measures for $p(TE | RE)$",
                         "The null hypothesis was rejected in Active Promoters,",
                         "with further weak evidence seen in Repressed Regions."),
         digits = c(0, 0, 0, 3, 3, 0),
         align = "llrrrl") %>%
  print(include.rownames = FALSE,
        caption.placement = "top")
@


<<lmTEgivRE>>=
lmTEgivRE <- fig2 %>%
  extract2("RE") %>%
  levels() %>%
  sapply(function(x){
    lm(logitTEgivRE ~ 0 + Region, data = filter(fig2, RE == x))
  },
  simplify = FALSE)
@


<<qqNormTEgivRE, echo = FALSE, fig.height=6, fig.cap='Normal QQ plots for the residuals from the conditional model for Regulatory Element.'>>=
par(mfrow = c(2, 3))
names(lmRETE) %>%
  lapply(function(x){
         plot(lmTEgivRE[[x]],
          which = 2,
          caption = NA,
          main = paste0(x, ":\n Normal Q-Q"))
  })
par(mfrow = c(1, 1))
@

\FloatBarrier
\subsection{Results of Analysis for Conditional Probabilities}
The test being used to detect any differing conditional probabilities across the genomic regions will again be Tukey's Honestly Significant Difference (HSD), which depends on the assumptions of normality and homoscedasticity performed above.
As also mentioned above, the sandwich estimator for the variance-covariance matrix was used in all pairwise comparisons to counteract any effects due to heteroscedasticity. \\
\\
Comparisons across regions will be performed separately for each RE, and the $p$-values for the complete set of $p$-values obtained will be adjusted for multiple comparisons using Bonferroni's methods to control the FWER across the entire experiment.
All results from pairwise comparisons are presented in Tables \ref{tab:TEgivREActivepromoter} to \ref{tab:TEgivREInsulator}.

<<resTEgivRE>>=
resTEgivRE <- names(lmTEgivRE) %>%
  lapply(function(x){
    glht(lmTEgivRE[[x]], linfct = mcp(Region = "Tukey"), vcov = sandwich) %>%
      summary(adjusted(type = "none")) %>%
      extract2("test") %>%
      extract(c("coefficients", "sigma", "tstat", "pvalues")) %>%
      as.data.frame %>%
      add_rownames("Comparison") %>%
      mutate(RE = x)}) %>%
  bind_rows %>%
  dplyr::select(RE, Comparison, 
                Estimate = coefficients,
                Std.Error = sigma,
                T = tstat,
                P = pvalues) %>%
  mutate(adjP = p.adjust(P, "bonferroni"),
         Sig = aster(adjP),
         Comparison = gsub("UTR(3|5)", "\\1'UTR", Comparison))
@

<<printResTEgivRE, echo=FALSE, include=TRUE, results='asis'>>=
xtabTEgivRE <- invisible(levels(fig2$RE) %>%
  lapply(function(x){
    resTEgivRE %>%
      filter(RE == x) %>%
      dplyr::select(Comparison, Estimate, 
                    `T-statistic` = T, 
                    `p-value` = adjP, 
                    ` ` = Sig) %>%
      xtable(label = gsub("_", "", paste0("tab:TEgivRE", x)),
             digits = c(0, 0, 3, 3, -2, 0),
             align = "llrrrl",
             caption = paste("Pairwise comparisons for probability of a base being a TE given it is within \\textbf{", 
                             capwords(gsub("_", " ", x)),
                             "Elements}.",
                             "Pairwise comparisons were performed using Tukey's HSD on the logit-transformed proportions of bases derived from the six sampled tissues.",
                             "The FWER was controlled using Bonferroni's adjustment on the set of $p$-values across all elements, and all $p$-values presented have been adjusted by this method.",
                             "A positive value in the Estimate column corresponds to a higher proportion in the first of the two genomic regions within the comparisons, whilst a negative value corresponds to a higher proportion in the second of the two regions.",
                             "Any significant values are indicated using asterisks as per the standard conventions of R.")) 
  }) %>%
    lapply(print,
           include.rownames = FALSE,
           caption.placement = "top",
           print.results = FALSE) %>%
  unlist %>%
  (function(x){gsub("(T|p)-", "$\\1$-",x)}))
cat(xtabTEgivRE)
@

\FloatBarrier
\subsection{Summary of Conditional Analysis}
Combining the data from Tables \ref{tab:TEgivREActivepromoter} to \ref{tab:TEgivREInsulator} and Figure \ref{fig:plotTEgivRE}:
\begin{itemize}
  \item The conclusion can be drawn that there is strong selection against TEs within REs located in CDS exons in comparison to every element in every genomic region.
  \item The selective pressure against TE seems to be slightly weaker in both promoter regions for CDS exons when compared to other element types.
  \item With the exception of Active Promoters, there is selective pressure against TEs located in the 3'UTR in comparison to all regions except CDS exons. 
  The lack of significance in Active Promoters is only in comparison to 5'UTRs, and this is likely to be a Type 2 error as the raw $p =$ \Sexpr{format(filter(resTEgivRE, grepl("3'UTR", Comparison), adjP > 0.05)$P, digits = 4, scientific = TRUE)}
  \item There is no evidence for any differing selective pressure acting on TEs between any of the REs located in 5'UTRs and CDS introns for any of the elements
  \item For Polycomb Repressed Regions and Insulators located in Intergenic regions, there was an increase in probability of being a TE when compared to CDS introns ($p$ = \Sexpr{format(filter(resTEgivRE, grepl("tron", Comparison), grepl("Intergenic", Comparison), RE == "Repressed_region")$adjP, digits = 3, scientific = TRUE)} \& $p$ = \Sexpr{format(filter(resTEgivRE, grepl("tron", Comparison), grepl("Intergenic", Comparison), RE == "Insulator")$adjP, digits = 3)} respectively), perhaps indicating even less selective pressure
  \item This pattern of an increased probability of being a TE  was also evident for Repressed Regions located in Intergenic and 5'UTRs ($p$ = \Sexpr{format(filter(resTEgivRE, grepl("5'UTR", Comparison), grepl("Intergenic", Comparison), RE == "Repressed_region")$adjP, digits = 3, scientific = TRUE)}).
\end{itemize}

\noindent These results also give rise to an alternate for Figure 2 in the submission to Heliyon (Figure \ref{fig:newPaperFig2}).\\
\\
We could also do a through comparison across elements, as there appears to be comparable selective pressure between Weak Promoters and Strong Enhancers, but strong pressure against TEs in Active Promoters in comparison to other RE types.
Weak Enhancers seem to show the the least amount of selective pressure, \textit{except} for those located in CDS exons.\\
\\
Additionally, as these measures are now independent a linear model could be applied across all elements \& regions.
The assumptions seem to hold, but this would rule out the new Figure 2 (below) as these fits were obtained for each element separately.


<<newPaperFig2, echo = FALSE, fig.height=4, fig.cap = 'A possible variation on Figure 2 as it could appear in the paper. The second row plots the conditional probabilities for all regions and all elements.'>>=
fs <- 8
c(lmRE, lmTEgivRE) %>% 
  lapply(summary) %>% 
  lapply(coef) %>%
  lapply(as.data.frame) %>%
  lapply(add_rownames, var = "Region") %>%
  bind_rows %>%
  mutate(RE = rep(rep(names(lmRE), each = 5), 2),
         RE = gsub("_", " ", RE),
         RE = capwords(RE),
         #RE = gsub(" ", "\n", RE),
         RE = factor(RE, levels = unique(RE)),
         Region = gsub("Region", "", Region),
         Region = gsub("UTR(3|5)", "\\1'UTR", Region),
         Region = gsub("CDS", "CDS ", Region),
         Region = factor(Region, levels = unique(Region)),
         Class = rep(c("RE in Genomic Region", "RE with TEs"), each = 30)) %>%
  mutate(pi = binomial()$linkinv(Estimate),
         lower = binomial()$linkinv(Estimate - `Std. Error`),
         upper = binomial()$linkinv(Estimate + `Std. Error`)) %>%
  ggplot(aes(x = Region, y = pi, fill = Region)) + 
  geom_bar(stat = "identity") +
  geom_errorbar(aes(ymin = lower, ymax = upper),
                width= 0.4) +
  theme_bw() +
  labs(x = "", y = expression(hat(pi))) +
  scale_y_continuous(expand=c(0.04,0)) + 
  facet_grid(Class~RE, scales = "free_y") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = 0, size = fs),
        axis.text.y = element_text(size = fs-1),
        strip.text = element_text(size = fs-1),
        legend.text = element_text(size = fs),
        legend.title = element_text(size = fs),
        panel.margin = unit(.004, "npc"),
        panel.grid = element_line(colour = "grey50")) 
@


 
\end{document}