---
title: "Figure 4"
author: "Steve Pederson"
date: "29 March 2016"
output: html_document
---
# Introduction
This contains code for the generation of Figure 4, S5 and S6

# Setup

```{r, loadPackages, message=FALSE, warning=FALSE}
library(ggplot2)
library(stringr)
library(reshape2)
library(matrixStats)
library(VennDiagram)
library(knitr)
library(xtable)
library(corrplot)
library(scales)
library(boot)
library(lmtest)
library(dplyr)
library(magrittr)
library(parallel)
library(funsForLu)
library(spShortcuts)
library(multcomp)
```

Load the data
```{r, loadData}
tpmFile <- file.path("data", "Figure4.txt")
tpmData <- read.delim(tpmFile, sep="\t", stringsAsFactors = FALSE) %>% 
  dplyr::select(ID, length, contains("TPM"), contains("IDs")) %>%
  tbl_df
```

Collect the elements by type and region
```{r, elements}
elements <- grep("IDs", colnames(tpmData), value = TRUE) %>% 
  str_split_fixed(pattern="_", 3) %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  dplyr::select(type = V1, region = V2) %>%
  lapply(unique) %>%
  lapply(sort)
nRegions <- length(elements$region)
nElements <- length(elements$type)
regFactor <- c("Proximal\nPromoter", "5'UTR", "CDS", "3'UTR")
```

Collect the tissues
```{r, tissues}
tissues <- colnames(tpmData) %>%
  str_subset("TPM") %>%
  str_replace("_TPM", "")
nTissues <- length(tissues)
```

Add columns of logical values to the main data object which denote the element type and region
```{r, addElements}
tpmData %<>%
  bind_cols(sapply(unlist(elements) %>% 
                     as.vector, 
                   function(x){dplyr::select(tpmData, contains(x)) %>% 
                                 rowSums %>% 
                                 as.logical},
                   simplify = FALSE)) %>%
  mutate(nElTypes = alu + L1 + L2 + LTR + mir)
```

Add further columns which convert the log(TPM) values to binary ```TRUE``` or ```FALSE``` variables for gene expression.
Genes for which log(TPM)$=-\infty$ have TPM$=0$ and will be set to ```FALSE```

```{r, addLogical}
tpmData %<>% 
  mutate(adipose_TF = as.logical(exp(adipose_TPM)),
         brain_TF = as.logical(exp(brain_TPM)),
         liver_TF = as.logical(exp(liver_TPM)),
         skeletal_muscle_TF = as.logical(exp(skeletal_muscle_TPM)),
         testes_TF = as.logical(exp(testes_TPM)),
         kidney_TF = as.logical(exp(kidney_TPM)))
```


# Setup For The Bootstrap Procedure

During the weighted bootstrap procedure, genes will be assigned to groups based on their TE structure.
In order to address any effects on TPM counts due to gene length, the TE groups will be further divided based on length quartiles.
The mean difference in length will also be recorded to ensure that the length distributions are matched.

Additionally, a number of genes contained all five elements, and as this set of `r tpmData %>% filter(alu, L1, L2, LTR, mir) %>% nrow()` genes was clearly higher than would be expected by random, this set with all 5 elements were omitted from the analysis.

```{r, makeBins}
nBins <- 10
tpmData %<>%
  filter(nElTypes != nElements) %>%
  mutate(lengthBin = findInterval(length, 
                                  quantile(length, 
                                           probs = seq(0, 1, length.out = nBins + 1)), 
                                  rightmost.closed=TRUE),
         logLength = log(length))
```

Confidence intervals will be formed for the difference in the probability of a gene being expressed, and for the difference in expression levels due to the presence of each TE within each Genomic Region.
However, in order to control the Type 1 Error at the level $\alpha = 0.05$, each set of $m$ intervals needs to be adjusted, with Bonferroni's method being the simplest approach.
Thus a set of $1-\alpha / m$ Confidence intervals will be formed for each genomic region.

```{r}
m <- nTissues*nElements*nRegions
alpha <- 0.05/m
nBoot <- 10000L
testVals <- c("logLength",
              paste0(tissues, "_TPM"),
              paste0(tissues, "_TF"))
```

In order to provide the number of samples to accurately construct these confidence intervals, the analysis was performed using `r nBoot` bootstrap iterations.

# The bootstrap procedure

## 3' UTR

### Alu Elements

```{r, aluUtr3Data}
aluUtr3Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("alu", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(alu_cds_IDs == 0, 
         alu_prox_IDs == 0, 
         alu_utr5_IDs == 0)
```

```{r, bootAluUtr3, cache = TRUE}
bootAluUtr3 <- mclapply(testVals, 
                        bootMeans, 
                        data = aluUtr3Data, 
                        testIds = filter(aluUtr3Data, alu)$ID, 
                        refIds = filter(aluUtr3Data, !alu)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootAluUtr3) <- testVals
```

### L1 Elements

```{r, L1Utr3Data}
L1Utr3Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L1", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L1_cds_IDs == 0, 
         L1_prox_IDs == 0, 
         L1_utr5_IDs == 0)
```

```{r, bootL1Utr3, cache = TRUE}
bootL1Utr3 <- mclapply(testVals, 
                        bootMeans, 
                        data = L1Utr3Data, 
                        testIds = filter(L1Utr3Data, L1)$ID, 
                        refIds = filter(L1Utr3Data, !L1)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootL1Utr3) <- testVals
```

### L2 Elements

```{r, L2Utr3Data}
L2Utr3Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L2", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L2_cds_IDs == 0, 
         L2_prox_IDs == 0, 
         L2_utr5_IDs == 0)
```

```{r, bootL2Utr3, cache = TRUE}
bootL2Utr3 <- mclapply(testVals, 
                        bootMeans, 
                        data = L2Utr3Data, 
                        testIds = filter(L2Utr3Data, L2)$ID, 
                        refIds = filter(L2Utr3Data, !L2)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootL2Utr3) <- testVals
```

### LTR Elements

```{r, LTRUtr3Data}
LTRUtr3Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("LTR", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(LTR_cds_IDs == 0, 
         LTR_prox_IDs == 0, 
         LTR_utr5_IDs == 0)
```

```{r, bootLTRUtr3, cache = TRUE}
bootLTRUtr3 <- mclapply(testVals, 
                        bootMeans, 
                        data = LTRUtr3Data, 
                        testIds = filter(LTRUtr3Data, LTR)$ID, 
                        refIds = filter(LTRUtr3Data, !LTR)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootLTRUtr3) <- testVals
```

### mir Elements

```{r, mirUtr3Data}
mirUtr3Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("mir", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(mir_cds_IDs == 0, 
         mir_prox_IDs == 0, 
         mir_utr5_IDs == 0)
```

```{r, bootMirUtr3, cache = TRUE}
bootMirUtr3 <- mclapply(testVals, 
                        bootMeans, 
                        data = mirUtr3Data, 
                        testIds = filter(mirUtr3Data, mir)$ID, 
                        refIds = filter(mirUtr3Data, !mir)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootMirUtr3) <- testVals
```

## 5' UTR

### Alu Elements

```{r, aluUtr5Data}
aluUtr5Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("alu", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(alu_cds_IDs == 0, 
         alu_prox_IDs == 0, 
         alu_utr3_IDs == 0)
```

```{r, bootAluUtr5, cache = TRUE}
bootAluUtr5 <- mclapply(testVals, 
                        bootMeans, 
                        data = aluUtr5Data, 
                        testIds = filter(aluUtr5Data, alu)$ID, 
                        refIds = filter(aluUtr5Data, !alu)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootAluUtr5) <- testVals
```

### L1 Elements

```{r, L1Utr5Data}
L1Utr5Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L1", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L1_cds_IDs == 0, 
         L1_prox_IDs == 0, 
         L1_utr3_IDs == 0)
```

```{r, bootL1Utr5, cache = TRUE}
bootL1Utr5 <- mclapply(testVals, 
                        bootMeans, 
                        data = L1Utr5Data, 
                        testIds = filter(L1Utr5Data, L1)$ID, 
                        refIds = filter(L1Utr5Data, !L1)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootL1Utr5) <- testVals
```

### L2 Elements

```{r, L2Utr5Data}
L2Utr5Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L2", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L2_cds_IDs == 0, 
         L2_prox_IDs == 0, 
         L2_utr3_IDs == 0)
```

```{r, bootL2Utr5, cache = TRUE}
bootL2Utr5 <- mclapply(testVals, 
                        bootMeans, 
                        data = L2Utr5Data, 
                        testIds = filter(L2Utr5Data, L2)$ID, 
                        refIds = filter(L2Utr5Data, !L2)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootL2Utr5) <- testVals
```

### LTR Elements

```{r, LTRUtr5Data}
LTRUtr5Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("LTR", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(LTR_cds_IDs == 0, 
         LTR_prox_IDs == 0, 
         LTR_utr3_IDs == 0)
```

```{r, bootLTRUtr5, cache = TRUE}
bootLTRUtr5 <- mclapply(testVals, 
                        bootMeans, 
                        data = LTRUtr5Data, 
                        testIds = filter(LTRUtr5Data, LTR)$ID, 
                        refIds = filter(LTRUtr5Data, !LTR)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootLTRUtr5) <- testVals
```

### mir Elements

```{r, mirUtr5Data}
mirUtr5Data <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("mir", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(mir_cds_IDs == 0, 
         mir_prox_IDs == 0, 
         mir_utr3_IDs == 0)
```

```{r, bootMirUtr5, cache = TRUE}
bootMirUtr5 <- mclapply(testVals, 
                        bootMeans, 
                        data = mirUtr5Data, 
                        testIds = filter(mirUtr5Data, mir)$ID, 
                        refIds = filter(mirUtr5Data, !mir)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootMirUtr5) <- testVals
```

## Proximal Promoter

### Alu Elements

```{r, aluProxData}
aluProxData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("alu", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(alu_cds_IDs == 0, 
         alu_utr3_IDs == 0, 
         alu_utr5_IDs == 0)
```

```{r, bootAluProx, cache = TRUE}
bootAluProx <- mclapply(testVals, 
                        bootMeans, 
                        data = aluProxData, 
                        testIds = filter(aluProxData, alu)$ID, 
                        refIds = filter(aluProxData, !alu)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootAluProx) <- testVals
```

### L1 Elements

```{r, L1ProxData}
L1ProxData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L1", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L1_cds_IDs == 0, 
         L1_utr3_IDs == 0, 
         L1_utr5_IDs == 0)
```

```{r, bootL1Prox, cache = TRUE}
bootL1Prox <- mclapply(testVals, 
                        bootMeans, 
                        data = L1ProxData, 
                        testIds = filter(L1ProxData, L1)$ID, 
                        refIds = filter(L1ProxData, !L1)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootL1Prox) <- testVals
```

### L2 Elements

```{r, L2ProxData}
L2ProxData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L2", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L2_cds_IDs == 0, 
         L2_utr3_IDs == 0, 
         L2_utr5_IDs == 0)
```

```{r, bootL2Prox, cache = TRUE}
bootL2Prox <- mclapply(testVals, 
                        bootMeans, 
                        data = L2ProxData, 
                        testIds = filter(L2ProxData, L2)$ID, 
                        refIds = filter(L2ProxData, !L2)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootL2Prox) <- testVals
```

### LTR Elements

```{r, LTRProxData}
LTRProxData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("LTR", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(LTR_cds_IDs == 0, 
         LTR_utr3_IDs == 0, 
         LTR_utr5_IDs == 0)
```

```{r, bootLTRProx, cache = TRUE}
bootLTRProx <- mclapply(testVals, 
                        bootMeans, 
                        data = LTRProxData, 
                        testIds = filter(LTRProxData, LTR)$ID, 
                        refIds = filter(LTRProxData, !LTR)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootLTRProx) <- testVals
```

### mir Elements

```{r, mirProxData}
mirProxData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("mir", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(mir_cds_IDs == 0, 
         mir_utr3_IDs == 0, 
         mir_utr5_IDs == 0)
```

```{r, bootMirProx, cache = TRUE}
bootMirProx <- mclapply(testVals, 
                        bootMeans, 
                        data = mirProxData, 
                        testIds = filter(mirProxData, mir)$ID, 
                        refIds = filter(mirProxData, !mir)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 1000L,
                        maxP = 0.01)
names(bootMirProx) <- testVals
```


## CDS regions

### Alu Elements

```{r, aluCDSData}
aluCDSData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("alu", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(alu_prox_IDs == 0, 
         alu_utr3_IDs == 0, 
         alu_utr5_IDs == 0)
```

```{r, bootAluCDS, cache = TRUE}
bootAluCDS <- mclapply(testVals, 
                        bootMeans, 
                        data = aluCDSData, 
                        testIds = filter(aluCDSData, alu)$ID, 
                        refIds = filter(aluCDSData, !alu)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 45L,
                        maxP = 0.1)
names(bootAluCDS) <- testVals
```

### L1 Elements

```{r, L1CDSData}
L1CDSData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L1", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L1_prox_IDs == 0, 
         L1_utr3_IDs == 0, 
         L1_utr5_IDs == 0)
```

```{r, bootL1CDS, cache = TRUE}
bootL1CDS <- mclapply(testVals, 
                        bootMeans, 
                        data = L1CDSData, 
                        testIds = filter(L1CDSData, L1)$ID, 
                        refIds = filter(L1CDSData, !L1)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 45L,
                        maxP = 0.1)
names(bootL1CDS) <- testVals
```

### L2 Elements

```{r, L2CDSData}
L2CDSData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("L2", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(L2_prox_IDs == 0, 
         L2_utr3_IDs == 0, 
         L2_utr5_IDs == 0)
```

```{r, bootL2CDS, cache = TRUE}
bootL2CDS <- mclapply(testVals, 
                        bootMeans, 
                        data = L2CDSData, 
                        testIds = filter(L2CDSData, L2)$ID, 
                        refIds = filter(L2CDSData, !L2)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 55L,
                        maxP = 0.1)
names(bootL2CDS) <- testVals
```

### LTR Elements

```{r, LTRCDSData}
LTRCDSData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("LTR", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(LTR_prox_IDs == 0, 
         LTR_utr3_IDs == 0, 
         LTR_utr5_IDs == 0)
```

```{r, bootLTRCDS, cache = TRUE}
bootLTRCDS <- mclapply(testVals, 
                        bootMeans, 
                        data = LTRCDSData, 
                        testIds = filter(LTRCDSData, LTR)$ID, 
                        refIds = filter(LTRCDSData, !LTR)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 33L,
                        maxP = 0.01)
names(bootLTRCDS) <- testVals
```

### mir Elements

```{r, mirCDSData}
mirCDSData <- tpmData %>%
  mutate(
    elBins = dplyr::select(
      tpmData, 
      one_of(grep("mir", elements$type, invert = TRUE, value = TRUE))
    ) %>%
    binsFromCols, 
    elBins = paste(elBins, lengthBin, sep = "_")) %>%
  filter(mir_prox_IDs == 0, 
         mir_utr3_IDs == 0, 
         mir_utr5_IDs == 0)
```

```{r, bootMirCDS, cache = TRUE}
bootMirCDS <- mclapply(testVals, 
                        bootMeans, 
                        data = mirCDSData, 
                        testIds = filter(mirCDSData, mir)$ID, 
                        refIds = filter(mirCDSData, !mir)$ID,
                        binCol = "elBins",
                        nBoot = nBoot, 
                        nGenes = 35L,
                        maxP = 0.1)
names(bootMirCDS) <- testVals
```

# Confidence Intervals for Length Differences

This step is a control to check that the length distributions were suitably paired.
Any intervals that don't contain zero will indicate that the two sets of genes will have different length distributions and the comparison will need to be reanalysed

```{r, allBoot}
allBoot<- list(bootAluUtr3, bootAluUtr5, bootAluProx, bootAluCDS,
               bootL1Utr3, bootL1Utr5, bootL1Prox, bootL1CDS,
               bootL2Utr3, bootL2Utr5, bootL2Prox, bootL2CDS,
               bootLTRUtr3, bootLTRUtr5, bootLTRProx, bootLTRCDS,
               bootMirUtr3, bootMirUtr5, bootMirProx, bootMirCDS) %>%
  set_names(elements$type %>%
              vapply(function(x){
                paste0(x, c("Utr3", "Utr5", "Prox", "CDS"))
              },
              character(4)) %>%
              as.vector())
```


```{r}
confIntLen <- allBoot %>%
  lapply(extract, "logLength") %>%
  lapply(getCI, alpha = 0.05, combined = TRUE) %>%
  bind_rows %>%
  mutate(Region = rep(c("3'UTR", "5'UTR","Proximal Promoter", "CDS"), 
                      times = nElements),
         TE = rep(elements$type, each = nRegions))
```

```{r, figS5, echo=FALSE, include=TRUE, fig.width=8}
figS5 <- confIntLen %>%
  mutate(TE = capwords(TE),
         TE = factor(TE, rev(unique(TE))),
         Region = gsub(" ", "\n", Region),
         Region = factor(Region, levels = regFactor)) %>%
  ggplot(aes(x = median, y = TE)) +
  geom_errorbarh(aes(xmin = lower, xmax = upper), height = 0.5) +
  facet_wrap(~Region, nrow=1, scales = "free_x") +
  geom_vline(xintercept = 0, colour = "grey50", linetype = 2) +
  theme_bw() +
  labs(x = "95% CI for the diference in log(Length)") 
figS5
```

As type II errors are the primary concern here, no adjustment was applied for multiple comparisons.
Zero didn't lie outside of any of these intervals, and as such the subsequent results for the bootstrap can be deemed to be valid.

```{r, exportFigS5, echo = FALSE, include=TRUE, message=FALSE}
png("FigureS5.png", height = 10, width = 14.5, units = "cm", res = 300)
fs <- 9
figS5 +
  theme(axis.text = element_text(size = fs - 0.5),
        axis.title = element_text(size = fs + 1),
        strip.text = element_text(size = fs + 1),
        panel.margin = unit(1, "mm"))
dev.off()
```


# Confidence Intervals for Elements & Regions

## Differences in the Probability of Gene Expression
```{r}
m <- nRegions*nElements*nTissues
confIntExProbs <- allBoot %>%
  lapply(extract, paste0(tissues, "_TF")) %>%
  lapply(getCI, 
         alpha = 0.05/m,
         combined = FALSE) %>%
  bind_rows() %>%
  mutate(TE = rep(elements$type, 
                  each = nRegions*nTissues),
         TE = capwords(TE),
         TE = gsub("Ltr", "LTR", TE),
         Region = c("3'UTR", "5'UTR", "Proximal\nPromoter", "CDS") %>%
           rep(times =nElements) %>%
           rep(each = nTissues),
         Region = factor(Region, levels= regFactor),
         Sig = sign(lower) == sign(upper),
         Tissue = capwords(Tissue, sep = "_"),
         Tissue = gsub(" Tf", "", Tissue),
         Tissue = factor(Tissue, rev(unique(sort(Tissue)))),
         lower = lower/log(2),
         median = median / log(2),
         upper = upper / log(2))
```


```{r, plotConfIntProbs, echo = FALSE, include = TRUE, fig.height=7, fig.width=10}
fig4a <- confIntExProbs %>%
  ggplot(aes(x = median, y = Tissue, colour = Sig)) +
  geom_point(size = 1)+
  geom_errorbarh(aes(xmin = lower,xmax =upper),
                 height = 0.5) +
  scale_color_manual(values = c("black", "red")) +
  geom_vline(xintercept = 0, linetype = 2, colour = "grey70", size = 0.5) +
  theme_bw() + 
  facet_grid(TE~Region, scales = "free_x") +
  guides(colour = FALSE) +
  labs(x = "Change in Pr(Expression)",
       y = c(),
       title = "Confidence Intervals for Change in Pr(Expression)")
fig4a
```


## Effects on Gene Expression

As there appeared to be a bias towards upregulatory effects due to the presence of an Alu element, tissue-specific effects were investigated for all elements

```{r, confIntTissues}
m <- nRegions*nElements*nTissues
confIntTissues <- allBoot %>%
  lapply(extract, paste0(tissues, "_TPM")) %>%
  lapply(getCI, 
         alpha = 0.05/m) %>%
  bind_rows() %>%
  mutate(TE = rep(elements$type, 
                  each = nRegions*nTissues),
         Region = c("3'UTR", "5'UTR", "Proximal\nPromoter", "CDS") %>%
           rep(times =nElements) %>%
           rep(each = nTissues),
         Sig = sign(lower) == sign(upper))
```

```{r, plotConfIntTissues, echo = FALSE, include=TRUE, fig.height=7, fig.width=10}
fig4b <- confIntTissues %>%
  mutate(Tissue = capwords(Tissue, sep = "_"),
         Tissue = gsub(" Tpm", "", Tissue),
         Tissue = factor(Tissue, rev(unique(sort(Tissue)))),
         TE = capwords(TE),
         TE = gsub("Ltr", "LTR", TE),
         Region = factor(Region, levels= regFactor),
         lower = lower/log(2),
         median = median / log(2),
         upper = upper / log(2)) %>%
  ggplot(aes(x = median, y = Tissue, colour = Sig)) +
  geom_point(size = 1)+
  geom_errorbarh(aes(xmin = lower,xmax =upper),
                 height = 0.5) +
  scale_color_manual(values = c("black", "red")) +
  geom_vline(xintercept = 0, linetype = 2, colour = "grey70", size = 0.5) +
  theme_bw() + 
  facet_grid(TE~Region, scales = "free_x") +
  guides(colour = FALSE) +
  labs(x = "FWER-adjusted 95% Confidence Intervals",
       y = c(),
       title = "Bootstrap Confidence Intervals For Changes In log(TPM)")
fig4b
```


```{r, confIntTissuesNoAdj}
confIntTissuesNoAdj <- allBoot %>%
  lapply(extract, paste0(tissues, "_TPM")) %>%
  lapply(getCI, 
         alpha = 0.05) %>%
  bind_rows() %>%
  mutate(TE = rep(elements$type, 
                  each = nRegions*nTissues),
         Region = c("3'UTR", "5'UTR", "Proximal\nPromoter", "CDS") %>%
           rep(times =nElements) %>%
           rep(each = nTissues),
         Sig = sign(lower) == sign(upper))
```

```{r, plotConfIntTissuesNoAdj, echo = FALSE, include=TRUE, fig.height=7, fig.width=10}
figS6 <- confIntTissuesNoAdj %>%
  mutate(Tissue = capwords(Tissue, sep = "_"),
         Tissue = gsub(" Tpm", "", Tissue),
         Tissue = factor(Tissue, rev(unique(sort(Tissue)))),
         TE = capwords(TE),
         TE = gsub("Ltr", "LTR", TE),
         Region = factor(Region, levels = regFactor),
         lower = lower/log(2),
         median = median / log(2),
         upper = upper / log(2)) %>%
  ggplot(aes(x = median, y = Tissue, colour = Sig)) +
  geom_point(size = 1)+
  geom_errorbarh(aes(xmin = lower,xmax =upper),
                 height = 0.5) +
  scale_color_manual(values = c("black", "red")) +
  geom_vline(xintercept = 0, linetype = 2, colour = "grey70", size = 0.5) +
  theme_bw() + 
  facet_grid(TE~Region, scales = "free_x") +
  guides(colour = FALSE) +
  labs(x = "Change in log(TPM)",
       y = c(),
       title = "Unadjusted 95% Confidence Intervals for Change in log(TPM)")
figS6
```


```{r, exportFigS6, echo=FALSE, include=FALSE}
png("FigureS6.png", height = 12, width = 14.5, units = "cm", res = 300)
fs <- 9
figS6 +
  labs(title = c(),
       x = expression(paste("Change in ", log[2],"(TPM)"))) +
  theme(axis.text = element_text(size = fs - 0.5),
        axis.title = element_text(size = fs + 1),
        strip.text = element_text(size = fs + 1),
        panel.margin = unit(1, "mm"))
dev.off()
```


```{r, exportFig4, echo=FALSE, include=FALSE}
png("Figure4.png", width = 14.5, height = 20, units = "cm", res = 300)
grid.newpage()
vp1 <- viewport(x = 0, y = 0.49, w = 1, h=0.51, just=c("left","bottom"))
vp2 <- viewport(x = 0, y = -0.01, w = 1, h = 0.51, just=c("left","bottom"))
print(fig4a + 
        labs(title = c(),
             x = "Difference in p(Expressed)") +
        theme(legend.title = element_text(size = fs),
              legend.text = element_text(size = fs ),
              legend.key.height = unit(0.025, "npc"),
              legend.position = "left",
              axis.text.y = element_text(size = fs - 1),
              axis.title.x = element_text(size = fs),
              strip.text.x = element_text(size = fs)),
      vp=vp1)
print(fig4b + 
        labs(title = c(),
             x = "Difference in Mean Expression") + 
        theme(axis.text = element_text(size = fs - 1),
              axis.title.x = element_text(size = fs),
              strip.text.x = element_text(size = fs),
              panel.margin = unit(0.005, "npc")),
      vp=vp2)
grid.text("A", x = 0.03, y = 0.97, default.units = "npc", gp = gpar(fontsize = fs + 4))
grid.text("B", x = 0.03, y = 0.47, default.units = "npc", gp = gpar(fontsize = fs + 4))
dev.off()
```

```{r}
save.image("Figure4.RData")
```

```{r}
sessionInfo()
```
