\documentclass{article}
\usepackage{underscore}
\usepackage{booktabs}
\usepackage{float}
\usepackage{subfig}
\usepackage{titling}
\usepackage{a4wide}
\usepackage{placeins}
\usepackage{amsmath}

\begin{document}

\title{Re-Analysis of Tissue Data}
\author{Steve Pederson\\
        Adelaide University\\
        Bioinformatics Hub}
\maketitle

\section{Required R Packages}

<<loadPackages, include=TRUE, echo=TRUE, warning=FALSE, message=FALSE>>=
library(ggplot2)
library(readr)
library(dplyr)
library(reshape2)
library(matrixStats)
library(xtable)
source("extraFunctions.R")
library(knitr)
@

<<knitrSetup, include=FALSE>>=
knit_hooks$set(crop=hook_pdfcrop)
opts_knit$set(self.contained=FALSE)
opts_chunk$set(fig.align='center', fig.show='as.is', crop=TRUE, 
               include=TRUE, echo=TRUE, 
               warning = FALSE, message = FALSE)
pdf.options(useDingbats=TRUE)
@

\section{The original FPKM data}

The original data, as represented by the FPKM measurement is briefly shown in this section.
Note the bi-modal distributions observed in this data (Figures \ref{fig:fpkmViolinPlots} \& \ref{fig:fpkmDensities}), with genes clearly observable as expressed, and not-expressed. \\

\noindent Two significant changes have been made in the current dataset.
\begin{enumerate}
  \item The main expression measure is now TPM, instead of FPKM
  \item The expression measure has not been log-transformed using the adjustment $\log (x +$ 1e-05).
  Undetectable genes will now have log-transformed expression values of $-\infty$ and can be simply removed from the analysis as non-expressed.
\end{enumerate}

As a result of these two changes, the bimodal distribution is no longer present in the data, and the analysis is greatly simplified.


<<fpkmData, cache=TRUE>>=
fpkmFile <- file.path("data", "vioplot_data")
fpkmData <- read_delim(fpkmFile, "\t")
fpkmData$RE <- fpkmData %>% 
  dplyr::select(contains("utr5")) %>% 
  rowSums %>% 
  as.logical
oldRE <- grep("utr5", colnames(fpkmData), value = TRUE)
@


<<fpkmViolinPlots, echo=FALSE, fig.width=6, fig.height=5, include=TRUE, fig.cap='FPKM: Violin plot for adipose tissue based on the presence or absence of a repeat element.'>>=
ggplot(fpkmData, aes(x=RE, y=adipose_expression)) + 
  geom_violin(fill="darkred") +
  geom_boxplot(width=0.1, fill="black", outlier.colour=NA) +
  stat_summary(fun.y=median, geom="point", fill="white", shape=21, size=5) +
  ylab("log2(FPKM)") + 
  xlab("Repeat Element") +
  scale_y_continuous(expand=c(0,0)) +
  ggtitle("Adipose Tissue") +
  theme(panel.background=element_rect("white"))
@

<<fpkmDensities, echo=FALSE, fig.cap='Expression densities for all tissues using FPKM data.', cache=TRUE, dependson='fpkmData'>>=
fpkmData %>% 
  dplyr::select(ID, contains("expression"), RE, gene_length) %>%
  melt(c("ID", "RE", "gene_length"), 
       grep("expression", colnames(fpkmData), value = TRUE),
       "Tissue",
       value.name = "FPKM") %>%
  mutate(Tissue = gsub("_expression", "", Tissue)) %>%
  mutate(Tissue = capwords(Tissue)) %>%
ggplot(aes(x = FPKM, fill = RE)) +
  geom_density(alpha=0.3) +
  facet_wrap(~Tissue, 3) +
  theme_bw() +
  labs(fill = "Repeat Element")
@

\clearpage
\section{TPM data}\label{sec:TPM}
<<TPMData, cache=TRUE>>=
tpmFile <- file.path("data", "new_whole_gene_expression_dataset.txt")
tpmData <- read.delim(tpmFile, sep="\t", stringsAsFactors = FALSE) %>% 
  select(ID, length, effective_length, contains("TPM"), one_of(oldRE)) %>%
  tbl_df
allTE <- grep("utr5", colnames(tpmData), value=TRUE)
tpmData$RE <- tpmData %>% 
  dplyr::select(one_of(oldRE)) %>%
  rowSums %>% 
  as.logical
@

<<tpmDensities, echo=FALSE, fig.width=7, fig.height=5, fig.cap='Expression densities for all tissues using TPM instead of FPKM data. Genes with repeat elements are shown in blue, whilst those without are shown in red.', cache=TRUE, dependson='TPMData'>>=
tpmData %>% 
  dplyr::select(ID, contains("TPM"), RE, length) %>%
  melt(c("ID", "RE", "length"), 
       grep("TPM", colnames(tpmData), value = TRUE),
       "Tissue",
       value.name = "TPM") %>%
  mutate(Tissue = gsub("_TPM", "", Tissue)) %>%
  mutate(Tissue = capwords(Tissue)) %>%
  filter(TPM > -Inf) %>%
ggplot(aes(x = TPM, fill = RE)) +
  geom_density(alpha=0.3) +
  facet_wrap(~Tissue, 3) +
  theme_bw() +
  labs(fill = "Repeat Element",
       x = "log(TPM)",
       y = "Density") +
  guides(fill = FALSE)
@

The new dataset contained annotations for multiple elements (e.g. utr3, prox etc) not included in the initial analysis \& these were removed for this particular analysis.
These can be re-incorporated later as required. \\

The previously mentioned changes to measurement of gene expression have completely removed the low-end noise seen in Figure \ref{fig:fpkmViolinPlots}, and as such the general densities as seen in Figure \ref{fig:tpmDensities} will be preferable to violin plots.
The patterns observed in Figure \ref{fig:tpmDensities} also show the clear differences in distributions for the expression levels for genes/transcripts containing an RE. \\

This change to the un-adjusted TPM \emph{completely removes the requirement for an Fisher Tests based on classification into low- or high- expressed gene groups}, and these results can now be removed from the submitted article.
This also means that any references to Odds Ratios can also be removed, further removing the point of confusion.
The following analysis is also much more straight forward. \\

In order to investigate whether length was having any effect on the presence of an RE, the distributions of gene length were compared across the two sets of genes for each tissue type.
Unsurprisingly, longer genes were more likely to have an RE (Figure \ref{fig:lengthDensities}).


<<lengthDensities, echo=FALSE, fig.width=4, fig.height=4, fig.cap='Gene length distributions. Genes with repeat elements are shown in blue, whilst those without are shown in red.', cache=TRUE, dependson='TPMData'>>=
tpmData %>% 
ggplot(aes(x = log(length), fill = RE)) +
  geom_density(alpha=0.3) +
  theme_bw() +
  labs(fill = "Repeat Element",
       x = "log(Length)",
       y = "Density") +
  guides(fill = FALSE)
@

% These are TTN & MUC16, which are the two longest genes
% These distort the diagnostic plots for fitting the linear model
% Checking the plots without them, was helpful, although they have been left in
% for the complete analysis
% TTN <- paste0("NM_", c("133378", "003319", "001267550", "133437", "024690"))

<<normalisedTPM, cache=TRUE, dependson='TPMData'>>=
normalisedTPM <- tpmData %>% 
  dplyr::select(ID, contains("TPM"), RE, length) %>%
  melt(c("ID", "RE", "length"), 
       grep("TPM", colnames(tpmData), value = TRUE),
       "Tissue",
       value.name = "TPM") %>%
  mutate(Tissue = gsub("_TPM", "", Tissue)) %>%
  mutate(Tissue = capwords(Tissue)) %>%
  filter(TPM > -Inf) %>%
  mutate(normalisedTPM = TPM - log(length)) %>%
  rename(Length = length) %>%
  tbl_df
@

It was also noted that shorter genes were more likely to have a higher TPM count, as seen in Figure \ref{fig:plotLengthVsTPM}.
Thus a normalised TPM value was created to give a TPM/nt value on the log scale:
\begin{align*}
  \text{normalisedTPM} = \log \left( \frac{\text{TPM}}{\text{length}} \right)
\end{align*}
This more effectively models the bias towards higher TPM counts for shorter genes, which can be considered to be more of a technical artefact for the data as opposed to true biology.

<<plotLengthVsTPM, fig.height=8, fig.cap="Raw TPM counts plotted against gene length.">>=
ggplot(normalisedTPM, aes(x=log(Length), y = TPM)) + 
  geom_point(size = 1.5, alpha = 0.3) + 
  theme_bw() +
  geom_smooth(se=FALSE) + 
  facet_wrap(~Tissue, ncol=2) +
  labs(x = "log (Gene Length)")
@

<<plotLengthVsnormalisedTPM, fig.height=8, fig.cap="normalised TPM counts plotted against gene length. Note the differing patterns for points associated with the longest gene in Skeletal Muscle. These represent transcripts of the gene TTN which is only expressed in muscle tissue.">>=
ggplot(normalisedTPM, aes(x=log(Length), y = normalisedTPM)) + 
  geom_point(size = 1.5, alpha = 0.3) + 
  theme_bw() +
  geom_smooth(se=FALSE) + 
  facet_wrap(~Tissue, ncol=2) +
  labs(x = "log (Gene Length)",
       y = "Normalised TPM")
@


<<tpmPerNt,echo = FALSE, fig.width=7, fig.height=5, include=TRUE, fig.cap='TPM values normalised to account for length. Genes with repeat elements are shown in blue, whilst those without are shown in red.'>>=
ggplot(normalisedTPM, aes(x = normalisedTPM, fill = RE)) +
  geom_density(alpha=0.3) +
  facet_wrap(~Tissue, 3) +
  theme_bw() +
  labs(fill = "Repeat Element",
       x = "log(TPM/Length)",
       y = "Density") +
  guides(fill = FALSE)
@


\noindent In all tissues, the normalised TPM values for genes with REs were clearly distributed below those of genes without REs.

<<lm>>=
lmTPM <- lm(normalisedTPM ~ 0 + Tissue + RE + Length, normalisedTPM)
@

<<tabRE, echo=FALSE, results='asis'>>=
summary(lmTPM) %>% 
  coef %>% 
  as.data.frame %>%
  add_rownames("Coefficient") %>%
  mutate(Coefficient = gsub("Tissue", "Tissue = ", Coefficient),
         Coefficient = gsub("RETRUE", "Repeat Element", Coefficient)) %>%
  xtable(caption = "Model Coefficients for the effect of the presence of repeat element across all tissues. No $p$-value adjustment was required due to the extreme $p$-values observed and the high level of significance.",
         label = "tab:lmTPM", 
         digits = c(0, 0, -2, -2, 2, -2)) %>%
  print(caption.placement = "top",
        include.rownames = FALSE)
@

As the plots in Figure \ref{fig:tpmPerNt} showed no clear violations of normality, a simple linear regression model was run across all tissues, including a term to account for the effects of gene length on the normalised TPM values.
The adjusted $R^2$ value for the model was \Sexpr{format(summary(lmTPM)$adj.r.squared, digits=4)} indicating a reasonable fit, and the coefficient for the presence of an RE was highly significant (Table \ref{tab:lmTPM}), indicating that on average, genes with an RE in the 5' UTR will have lower gene expression levels, as measured using the above values. \\

The above model was also fit (data not shown) incorporating an interaction term for the presence of an RE with the Tissue type, in order to assess whether any tissue specific impacts were detectable.
None were detected (all raw $p$-values $> 0.01$).

\clearpage
\appendix


\section{Additional R Info}

The above required the additional functions from the file \texttt{extraFunctions.R} as provided below

<<printCapwords, echo=TRUE, include=TRUE, tidy=TRUE, comment="", results='markup'>>=
capwords
@


<<sessionInfo, echo=FALSE, include=TRUE, results='markup'>>=
sessionInfo()
@



\end{document}